File: c/1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#include <time.h>
#define defaultInput "../../Inputs/2020_1.txt"
#define dataLine 20

int findLines(char *fileName) {
	int numLines = 0;
	char textRead[dataLine];

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return -1;
    }

	while(fgets(textRead, dataLine, inFile)) {
        ++numLines;
	}

	fclose(inFile);

	return numLines;
}

void readData(int *nums, char* fileName) {
    int lineNum = 0;
	char textRead[dataLine];

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return;
    }

	while(fgets(textRead, dataLine, inFile)) {
        nums[lineNum] = atoi(textRead);

        lineNum++;
	}

	fclose(inFile);

    return;
}

int part1(int *nums, int numLines) {
    int product;

    for (int i = 0; i < numLines - 1; i++) {
        for (int j = i + 1; j < numLines; j++) {
            if (nums[i] + nums[j] == 2020) {
                product = nums[i] * nums[j];
            }
        }
    }

    return product;

}

int part2(int *nums, int numLines) {
    int product;

    for (int i = 0; i < numLines - 2; i++) {
        for (int j = i + 1; j < numLines - 1; j++) {
            for (int k = j + 1; k < numLines; k++) {
                if (nums[i] + nums[j] + nums[k] == 2020) {
                    product = nums[i] * nums[j] * nums[k];
                }
            }
        }
    }

    return product;
}

int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    int numLines, product;
	numLines = findLines(inputPath);

	if (numLines == -1) {
        printf("Error: Could not read input file. Quitting\n");
        return -1;
	}

	int nums[numLines];

    readData(&nums[0], inputPath);

    clock_t t; 
    t = clock(); 
    int p1 = part1(&nums[0], numLines);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nProduct of numbers that sum to 2020: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock();
    int p2 = part2(&nums[0], numLines);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nProduct of numbers that sum to 2020: %d\nRan in %f seconds\n", p2, t_p2);

	return 0;
}

File: python/1.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
def part1(data):
    """ 2020 Day 1 Part 1

    >>> part1(['1721', '979', '366', '299', '675', '1456'])
    514579
    """

    nums = sorted([int(line) for line in data])

    found = False
    for i in range(len(nums) - 1):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == 2020:
                found = True
                break

        if found:
            break

    return nums[i] * nums[j]


def part2(data):
    """ 2020 Day 1 Part 2

    >>> part2(['1721', '979', '366', '299', '675', '1456'])
    241861950
    """

    nums = sorted([int(line) for line in data])

    found = False
    for i in range(len(nums) - 2):
        for j in range(i + 1, len(nums) - 1):
            for k in range(j + 1, len(nums)):
                if nums[i] + nums[j] + nums[k] == 2020:
                    found = True
                    break

            if found:
                break

        if found:
            break

    return nums[i] * nums[j] * nums[k]


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nProduct of numbers that sum to 2020: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nProduct of numbers that sum to 2020: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)